File: packages/api/src/auth/keys.js
const crypto = require("crypto");
const { signAccessToken, generateRefreshToken } = require("./jwt");

/**
 * Single-admin API key authentication.
 * The API_KEY env var is compared via constant-time hash comparison.
 */

function hashKey(key) {
  return crypto.createHash("sha256").update(key).digest("hex");
}

/** Validate an API key and return tokens */
function authenticateWithKey(providedKey) {
  const expectedKey = process.env.API_KEY;
  if (!expectedKey) {
    throw new Error("API_KEY not configured");
  }

  const providedHash = hashKey(providedKey);
  const expectedHash = hashKey(expectedKey);

  if (!crypto.timingSafeEqual(Buffer.from(providedHash), Buffer.from(expectedHash))) {
    return null;
  }

  const userId = "admin";
  const role = "admin";

  const accessToken = signAccessToken({ sub: userId, role });
  const refreshToken = generateRefreshToken(userId, role);

  return { accessToken, refreshToken, userId, role };
}

module.exports = { authenticateWithKey };

File: packages/windows-agent/auth/keys.py
import hashlib
import hmac
from config import API_KEY
from auth.jwt_auth import sign_access_token, generate_refresh_token


def authenticate_with_key(provided_key):
    if not API_KEY:
        raise ValueError("API_KEY not configured")

    provided_hash = hashlib.sha256(provided_key.encode()).digest()
    expected_hash = hashlib.sha256(API_KEY.encode()).digest()

    if not hmac.compare_digest(provided_hash, expected_hash):
        return None

    user_id = "admin"
    role = "admin"
    access_token = sign_access_token({"sub": user_id, "role": role})
    refresh_token = generate_refresh_token(user_id, role)

    return {
        "accessToken": access_token,
        "refreshToken": refresh_token,
        "userId": user_id,
        "role": role,
    }
